crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -468,7 +476,7 @@ fn normalize_string(input: &str, quotes: StringQuotes) -> (Cow<str>, ContainsNew
         } else if c == '\n' {
             newlines = ContainsNewlines::Yes;
         } else if !quotes.triple {
-            if c == '\\' {
+            if !is_raw && c == '\\' {
```

Comment:
Do we need to move this check to line 490 instead because we need to make sure that quotes are properly escaped. Can you add the following test

```python
r'It\'s normalizing \' and " quotes'
```

This should be formatted as:

```python
r"It's normalizing ' and \" quotes"r
```

Reply:
`r'It\'s normalizing \' and " quotes'` is not equivalent to `r"It's normalizing ' and \" quotes"`:

```python
>>> r'It\'s normalizing \' and " quotes'
'It\\\'s normalizing \\\' and " quotes'
>>> r"It's normalizing ' and \" quotes"
'It\'s normalizing \' and \\" quotes'
>>> r'It\'s normalizing \' and " quotes' == r"It's normalizing ' and \" quotes"
False
```

Reply:
Okay, I now played with your PR and found an example that produces invalid syntax:

```python
# Input
r'Not-so-tricky "quote \'\''

# Ruff
r"Not-so-tricky "quote \'\'"

# Black
r'Not-so-tricky "quote \'\''
```

Note how Ruff changes the quotes from `'` to `"` but fails to escape the `"`.  

We need to play a bit more with black to understand how black determines the preferred quotes for raw strings, and how the normalization has to work. Maybe @konstin knows more, because I'm not that familiar with Python and I must say, the escaping logic behind raw strings is confusing to me (you have to escape quotes)

Reply:
Thanks for the investigation. I'm reading the black source code. It looks like black returns the original string if it contains unescaped quotes:

https://github.com/psf/black/blob/1a972e3e11b144912155babdf48ff23d68059d57/src/black/strings.py#L201

Reply:
tbh i find python's raw string escaping rules confusing and i think there are cases that are just not properly representable (as the case above where black returns)